src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -0,0 +1,266 @@
+  if (_resizable && number_of_entries() > (_resize_load_trigger*table_size())) {
+    int desired_size = calculate_resize(true);
+    if (desired_size == table_size()) {
+      _resizable = false; // hit max
+    } else {
+      if (!resize(desired_size)) {
+        // Something went wrong, turn resizing off
+        _resizable = false;
+      }
+      log_info(jvmti, table) ("JvmtiTagMap table resized to %d", table_size());
+    }
+  }
+}
+
+// Serially remove unused oops from the table, and notify jvmti.
+void JvmtiTagMapTable::unlink_and_post(JvmtiEnv* env) {
+
+  // Does this environment have the OBJECT_FREE event enabled?
+  bool post_object_free = env->is_enabled(JVMTI_EVENT_OBJECT_FREE);
+
+  int oops_removed = 0;
+  int oops_counted = 0;
+  for (int i = 0; i < table_size(); ++i) {
+    JvmtiTagMapEntry** p = bucket_addr(i);
+    JvmtiTagMapEntry* entry = bucket(i);
+    while (entry != NULL) {
+      oops_counted++;
+      oop l = entry->object_no_keepalive();
+      if (l != NULL) {
+        p = entry->next_addr();
+      } else {
+        // Entry has been removed.
+        oops_removed++;
+        log_trace(jvmti, table)("JvmtiTagMap entry removed for index %d", i);
+        jlong tag = entry->tag();
+        *p = entry->next();
+        free_entry(entry);
+
+        // post the event to the profiler
+        if (post_object_free) {
+          JvmtiExport::post_object_free(env, tag);
+        }
+
+      }
+      // get next entry
+      entry = (JvmtiTagMapEntry*)HashtableEntry<WeakHandle, mtServiceability>::make_ptr(*p);
+    }
+  }
+
+  log_info(jvmti, table) ("JvmtiTagMap entries counted %d removed %d; %s",
+                          oops_counted, oops_removed, post_object_free ? "free object posted" : "no posting");
+}
+
+// Rehash oops in the table
+void JvmtiTagMapTable::rehash() {
+
+  ResourceMark rm;
+  GrowableArray<JvmtiTagMapEntry*> moved_entries;
+
+  int oops_counted = 0;
+  for (int i = 0; i < table_size(); ++i) {
+    JvmtiTagMapEntry** p = bucket_addr(i);
+    JvmtiTagMapEntry* entry = bucket(i);
+    while (entry != NULL) {
+      oops_counted++;
+      oop l = entry->object_no_keepalive();
+      if (l != NULL) {
+        // Check if oop has moved, ie its hashcode is different
+        // than the one entered in the table.
+        unsigned int new_hash = compute_hash(l);
+        if (entry->hash() != new_hash) {
+          *p = entry->next();
+          entry->set_hash(new_hash);
+          unlink_entry(entry);
+          moved_entries.push(entry);
+        } else {
+          p = entry->next_addr();
+        }
+      } else {
+        // Skip removed oops. They may still have to be posted.
+        p = entry->next_addr();
+      }
+      // get next entry
+      entry = (JvmtiTagMapEntry*)HashtableEntry<WeakHandle, mtServiceability>::make_ptr(*p);
+    }
+  }
+
+  int rehash_len = moved_entries.length();
+  // Now add back in the entries that were removed.
+  for (int i = 0; i < moved_entries.length(); i++) {
```

Comment:
rehash_len is read, but not used in for loop condition.

Reply:
It's in the logging message below.  I'll use it in the loop too.